### PR TITLE
 Support Libreswan "cable" creation

### PR DESCRIPTION
**Pull request by skitt**
This relies on the whack command-line tool for now. Because the
full-blown status is very hard to parse, the connection-tracking code
uses the simpler trafficstatus output, which doesn’t give all the
information we’d really need but is sufficient to determine whether a
connection is active or not.

We set up one tunnel per subnet pair, one tunnel between the hosts,
and one tunnel between each host and each opposing subnet.

Fixes: #592
Fixes: #385
Fixes: #564
Signed-off-by: Stephen Kitt <skitt@redhat.com>

### CODE REVIEW

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -16,10 +16,16 @@ jobs:
...
+        cable_driver: ['', 'wireguard', 'libreswan']
         exclude:
+# Our Helm setup doesn’t know how to deploy other cable drivers
           - deploytool: 'helm'
             cable_driver: 'wireguard'
+          - deploytool: 'helm'
+            cable_driver: 'libreswan'
+# TODO: Remove exclusion when issue #552 is fixed
+          - globalnet: '--globalnet'
+            cable_driver: 'wireguard'
```
Comment (mangelajo) [inline .github/workflows/e2e.yaml:28]:
```suggestion
```
The issue was fixed by:
    https://github.com/submariner-io/submariner/pull/562

Now we have an all-together different issue because the azure-ubuntu image has broken headers for wireguard... :/ https://github.com/submariner-io/submariner/issues/607

But I think it's unrelated to your patch, let's disable globally, we can ignore those runs for your patch.

File: pkg/cable/libreswan/libreswan.go [hunk 2]
```diff
@@ -1,57 +1,377 @@
...
+
+				/*
+					if err := Route(endpoint.Spec.CableName); err != nil {
+						return "", fmt.Errorf("Error routing connection %v: %v", endpoint.Spec.CableName, err)
+					}
+
+					if err := Initiate(endpoint.Spec.CableName); err != nil {
+						return "", fmt.Errorf("Error initiating connection %v: %v", endpoint.Spec.CableName, err)
+					}
+				*/
```
Comment (mangelajo) [inline pkg/cable/libreswan/libreswan.go:257]:
```suggestion
```

File: pkg/cable/libreswan/libreswan.go [hunk 3]
```diff
@@ -1,57 +1,377 @@
...
+					//return "", fmt
+					klog.Errorf("Error routing connection %s: %v", connectionName, err)
+				}
+
+				cmd = exec.Command("/usr/libexec/ipsec/whack", "--initiate", "--name", connectionName)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+
+				if err := cmd.Run(); err != nil {
+					//return "", fmt
```
Comment (mangelajo) [inline pkg/cable/libreswan/libreswan.go:273]:
```suggestion
```

File: pkg/cable/libreswan/libreswan.go [hunk 4]
```diff
@@ -1,57 +1,377 @@
...
+						return fmt.Errorf("Error deleting a connection with args %v: %v", args, err)
+					}
+				}
+			}
+		}
+	}
+
+	for j := range i.connections {
+		if i.connections[j].Endpoint.CableName == endpoint.Spec.CableName {
+			if j == 0 && j == len(i.connections)-1 {
```
Comment (mangelajo) [inline pkg/cable/libreswan/libreswan.go:322]: how can j be two different things at once? I guess for len(i.connections)==1 

nit: can we use j==0 && len(i.connections)==1 ?

File: pkg/cable/libreswan/libreswan.go [hunk 5]
```diff
@@ -1,57 +1,377 @@
...
+						klog.Errorf("Error deleting a connection with args %v; got exit code %d: %v", args, err.ExitCode(), err)
+					default:
+						return fmt.Errorf("Error deleting a connection with args %v: %v", args, err)
+					}
+				}
+			}
+		}
+	}
+
+	for j := range i.connections {
```
Comment (mangelajo) [inline pkg/cable/libreswan/libreswan.go:320]: can we refactor it out to a separate method? (removeConnectionForEndpoint?)

File: pkg/cable/libreswan/libreswan.go (hunk 4 above)
Comment (skitt) [reply pkg/cable/libreswan/libreswan.go:322]: I prefer the current form because it matches the other checks...

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (skitt) [reply .github/workflows/e2e.yaml]: Done, I didn’t notice that that part had gone.

File: pkg/cable/libreswan/libreswan.go (hunk 2 above)
Comment (skitt) [reply pkg/cable/libreswan/libreswan.go]: Done.

File: pkg/cable/libreswan/libreswan.go (hunk 5 above)
Comment (skitt) [reply pkg/cable/libreswan/libreswan.go]: Done.

File: pkg/cable/libreswan/libreswan.go (hunk 4 above)
Comment (mangelajo) [reply pkg/cable/libreswan/libreswan.go]: ack :+1:

File: pkg/cable/libreswan/libreswan.go [hunk 6]
```diff
@@ -1,57 +1,368 @@
...
 
 // Init initializes the driver with any state it needs.
-func (i *libreSwan) Init() error {
+func (i *libreswan) Init() error {
+	// Write the secrets file:
+	// %any %any : PSK "secret"
+	// TODO Check whether the file already exists
+	file, err := os.Create("/etc/ipsec.d/submariner.secrets")
+	if err != nil {
+		return fmt.Errorf("Error creating the secrets file: %v", err)
```
Comment (sridhargaddam) [inline pkg/cable/libreswan/libreswan.go:87]: nit: When returning an error message in an API, it's recommended to use small letter (i.e., instead of Error, please use error)

File: pkg/cable/libreswan/libreswan.go [hunk 7]
```diff
@@ -1,57 +1,368 @@
...
+	}
+	defer file.Close()
+
+	fmt.Fprintf(file, "%%any %%any : PSK \"%s\"\n", i.secretKey)
+
+	// Ensure Pluto is started
+	if err := i.runPluto(); err != nil {
+		return fmt.Errorf("Error starting Pluto: %v", err)
```
Comment (sridhargaddam) [inline pkg/cable/libreswan/libreswan.go:95]: nit: as above in all occurrences.

File: pkg/cable/libreswan/libreswan.go (hunk 6 above)
Comment (skitt) [reply pkg/cable/libreswan/libreswan.go]: Done, thanks for the reminder!
